### PR TITLE
fix(router): support lazy loading for named outlets

### PR DESCRIPTION
**Pull request by asvishnyakov**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

**This is correct implementation and replacement of #25483**

This PR is avoid problem appear in #25483: the same error occurs when you have lazy-loaded default route and routes with empty path and named outlet before it.

---

Given the following route configuration:

```
[
    {path: '', loadChildren: 'children'},
    {path: 'a', loadChildren: 'otherChildren', outlet: 'aux'}
]
```
With this configuration, using the outlet as an entry point to the application, the root module will not load properly and results in the following error:

`TypeError: Cannot read property 'routes' of undefined at getChildConfig (router.js:3041)`

The `_loadedConfig` property of the route remains `undefined` as the `expandSegmentAgainstRoute()` isn't executed.

This issue is only present if the root has a lazy loaded module, works fine with a component or even with a lazy loaded module if the root path is not empty.

**Issue Number:** #12842

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by asvishnyakov** [comment]
Also, possible fix angular/angular#10981
@Jonathan002 FYI
@jasonaden May you review?

**Comment by asvishnyakov** [comment]
Any update?

**Comment by Flyrell** [comment]
@asvishnyakov Thank you for this, I hope Angular team will take their time to review those few lines of code before v9 is released (although, it might be too much to ask).

Anyways, there are some indentation issues in `packages/router/src/apply_redirects.ts` which might make things go a bit slower to accept and merge.

**Comment by dustintheweb** [comment]
This is still not working fully & https://github.com/angular/angular/issues/10981#issuecomment-301787482 is locked for some reason. Plz halp.

File: packages/router/src/apply_redirects.ts [hunk 1]
```diff
@@ -144,28 +144,29 @@ class ApplyRedirects {
...
-        concatAll(), first((s: any) => !!s), catchError((e: any, _: any) => {
-          if (e instanceof EmptyError || e.name === 'EmptyError') {
-            if (this.noLeftoversInUrl(segmentGroup, segments, outlet)) {
-              return of (new UrlSegmentGroup([], {}));
-            }
-            throw new NoMatch(segmentGroup);
-          }
-          throw e;
-        }));
+    return combineLatest(routes.map((r: any) => {
```
Comment (atscott) [inline packages/router/src/apply_redirects.ts]: Is this change necessary? The code may have been updated since this PR was created. Your test passes if I revert these lines.

**Comment by khiami** [comment]
Any updates on this ticket?

**Comment by atscott** [comment]
* Rebased PR
* Updated `routes.map` to use `try...catch` and `throwError` to ensure possible errors thrown from `expandSegmentAgainstRoute` are handled by `catchError` code
* Updated test to more accurately reflect the use-case in the issue report and so that it would fail without the changes in `apply_redirects`

The internal tests have quite a few failures so this will require further investigation into why this change breaks them.

**Comment by atscott** [comment]
@googlebot I consent

**Comment by atscott** [comment]
I reviewed one of the failing tests and found that it fails because this finds all configs which match the url tree and loads each one if they have lazy children. The previous behavior would find and load only the first match. Generally, this would be fine because the router _usually_ only activates a single config. Matches that appear earlier in the config take precedence (which is working as intended and documented). The failing provides an override to the real route so that it loads a no-op module. This change makes it so that the test override _and_ the real `Route` match and are loaded, resulting in the failure.

i.e.
```
const REAL_ROUTES = [
  {
    path: 'b/c',
    loadChildren: 'xyz',
    pathMatch: 'full'
  },
];
=====

...
{
  testRoutes: [
  routes: [
    {path: 'a/b/c', loadChildren: 'noop_module'},
    {path: 'a', children: REAL_ROUTES},
  ],
}
```

In our case with these empty path named outlets, we have to load matched configs from each outlet. This approach also presents a problem because it will always load wildcard routes, which will always match the given url tree. At the moment, I can't really think of a solution to this. If no backwards compatible solution can be found, we could provide this as opt-in behavior in `ExtraOptions`, something like `lazyRouteLoadingBehavior: 'firstMatch'|'allMatches'`

**Comment by atscott** [comment]
Closing in favor of #38379, as that one addresses the issues mentioned in https://github.com/angular/angular/pull/32688#issuecomment-669623926
